### PR TITLE
feat: Add cache for github actions

### PR DESCRIPTION
**Pull request by dhruvmanila**
Added the ability to cache the pip downloads. The key is based on the OS which the tests are run on and the `requirements.txt` file. This means that if either the OS changes or the `requirements.txt` is modified, the cache will be updated automatically.

### CODE REVIEW

**Comment by cclauss** [comment]
Nice!  Is there a way to provide before/after numbers to demonstrate how much time this saves?

**Comment by dhruvmanila** [comment]
Yes, I will push an empty commit right now to compare the numbers.

**Comment by dhruvmanila** [comment]
No a major difference but the pip downloads are cached for future use:
- Fresh download: https://github.com/internetarchive/infogami/runs/2284549247#step:6:8
- Using cached: https://github.com/internetarchive/infogami/pull/161/checks?check_run_id=2284579116#step:6:8

**Comment by anandology** [comment]
using pip cache doesn't give much speed up as it only saves the download time and not the time it takes to install the library.  What i've tried in one of my projects is to cache the virtualenv used to run the tests. That have very good speedup. 

Here is what I've used:

```
    - uses: actions/cache@v1
      with:
        path: ~/venv
        key: ${{ runner.os }}-venv-${{ hashFiles('*requirements.txt') }}
        restore-keys: |
          ${{ runner.os }}-venv-
    - name: setup virtualenv
      run: |
        python -m venv ~/venv
    - name: Install dependencies
      run: |
        source ~/venv/bin/activate
        pip install -r dev-requirements.txt -r requirements.txt
```

You need to make sure to activate venv before running the tests. 

@dhruvmanila can you try with this approach and see if this gives a better speedup?

**Comment by dhruvmanila** [comment]
Hey, thanks for the insight. That is a really good idea. I will try it out now.

File: .github/workflows/python_tests.yml [hunk 1]
```diff
@@ -35,6 +35,12 @@ jobs:
       - uses: actions/setup-python@v2
         with:
           python-version: ${{ matrix.python-version }}
+      - uses: actions/cache@v2
+        with:
+          path: ~/infogami_python_venv
+          key: ${{ runner.os }}-venv-${{ hashFiles('*requirements.txt') }}
+      - run: python3 -m venv ~/infogami_python_venv
+      - run: source ~/infogami_python_venv/bin/activate
       - run: pip install black codespell flake8 isort mypy pytest pyupgrade
```
Comment (anandology) [inline .github/workflows/python_tests.yml:44]: You need to source venv/bin/activate for every step. Each step is completely independent and the environment of a step is not passed to the next step. 

```
    - run: |
        source ~/infogami_python_venv/bin/activate
        pip install black codespell flake8 isort mypy pytest pyupgrade
    - run: |
        source ~/infogami_python_venv/bin/activate
        black --check . || true
```

Comment (anandology) [reply .github/workflows/python_tests.yml:44]: It may also be a good idea to add a dev-requirements.txt file.

**Review by anandology — CHANGES_REQUESTED** [review]

File: .github/workflows/python_tests.yml (hunk 1 above)
Comment (dhruvmanila) [reply .github/workflows/python_tests.yml]: Yes, that and I am also thinking of grouping all the `run` steps into a bash script in the `scripts/` directory to keep the workflow file simple.

**Comment by dhruvmanila** [comment]
Yes, it is working now: https://github.com/internetarchive/infogami/runs/2285449631#step:7:10

Next steps:
- [ ] Put all the test dependencies in `requirements_test.txt` file
- [ ] Put all the steps in `scripts/workflow_python_test.sh` file

**Comment by anandology** [comment]
Cool. Did you measure the time improvement?

**Comment by cclauss** [comment]
Nice progress.  I usually use separate run statements so we can quickly see which test is failing.  

This is currently difficult with steps like black and flake8 writing a ton to the logs but will soon calm down when the code becomes black compliant.

**Comment by dhruvmanila** [comment]
@cclauss Should I go ahead and separate out the run steps in a bash script as mentioned? This would make the workflow file a bit simpler to manage with all the indentation.

@anandology 
[Time difference](https://github.com/internetarchive/infogami/actions/workflows/python_tests.yml): `1m 57s -> 1m 29s`
Sample: 1 (more sample runs are required to get a proper conclusion)

**Comment by anandology** [comment]
On Wed, Apr 7, 2021 at 1:42 PM Christian Clauss ***@***.***>
wrote:

> Nice progress. I usually use separate run statements so we can quickly see
> which test is failing.
>
> This is currently difficult with steps like black and flake8 writing a ton
> to the logs but will soon calm down when the code becomes black compliant.
>
+1.
It is good to keep the steps atomic.

**Comment by cclauss** [comment]
> pip install -r dev-requirements.txt -r requirements.txt

Let's `git mv dev-requirements.txt requirements_test.txt` to match Open Library's naming conventions.

The test requirements assume that the runtime requirements are installed so to reduce the chances of errors, let's make the first line of `requirements_test.txt` import `requirements.txt` as is done in https://github.com/internetarchive/openlibrary/blob/master/requirements_test.txt

File: .github/workflows/python_tests.yml [hunk 2]
```diff
@@ -38,28 +38,15 @@ jobs:
       - uses: actions/cache@v2
         with:
           path: ~/infogami_python_venv
-          key: ${{ runner.os }}-venv1-${{ hashFiles('*requirements.txt') }}
+          key: ${{ runner.os }}-venv-${{ hashFiles('requirements.txt') }}-${{ hashFiles('requirements_test.txt') }}
```
Comment (cclauss) [inline .github/workflows/python_tests.yml:41]:
```suggestion
          key: ${{ runner.os }}-venv-${{ hashFiles('requirements*.txt') }}
```
Would that work?

File: requirements_test.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+-r requirements.txt
+black
+codespell
+flake8
+isort
+mypy
+pytest
+pyupgrade
+psycopg2
```
Comment (cclauss) [inline requirements_test.txt:9]: @anandology Should `psycopg2` be moved to `requirements.txt` or is Postgres optional at runtime?

File: .github/workflows/python_tests.yml (hunk 2 above)
Comment (dhruvmanila) [reply .github/workflows/python_tests.yml:41]: I don't think so that's how globs work, but we could try. I am happy to be wrong about that :)

**Comment by dhruvmanila** [comment]
Running `ls` locally with this pattern does work:
```console
❯ ls requirements*
requirements.txt      requirements_test.txt
❯ ls requirements*.txt
requirements.txt      requirements_test.txt
```

We can perform the following steps to check whether it is working on GitHub actions:
1. Empty commit to trigger workflow to check whether it is cached
2. Add random dependency (`cowsay`) to `requirements_test.txt`
3. Empty commit to trigger workflow to check whether it is cached
4. Remove the random dependency

**Review by cclauss — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**